tests/unit/test_hnsw_tiered.cpp

```diff
@@ -1905,3 +1907,757 @@ TYPED_TEST(HNSWTieredIndexTest, alternateInsertDeleteAsync) {
+            .jobQueueCtx = &index_ctx,
+            .submitCb = submit_callback,
+            .memoryCtx = &memory_ctx,
+            .UpdateMemCb = update_mem_callback,
+            .primaryIndexParams = &params,
+        };
+        auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+            TieredFactory::NewIndex(&tiered_params, allocator));
+        // Set the created tiered index in the index external context.
+        index_ctx.index_strong_ref.reset(tiered_index);
+        EXPECT_EQ(index_ctx.index_strong_ref.use_count(), 1) << decider_name;
+
+        auto *hnsw = tiered_index->index;
+        auto *flat = tiered_index->flatBuffer;
+
+        // For every i, add the vector (i,i,i,i) under the label i.
+        for (size_t i = 0; i < n; i++) {
+            auto cur = decider(i, n) ? hnsw : flat;
+            GenerateAndAddVector<TEST_DATA_T>(cur, d, n_labels - (i % n_labels), i);
+        }
+        ASSERT_EQ(VecSimIndex_IndexSize(tiered_index), n) << decider_name;
+
+        // Query for (n,n,n,n) vector (recall that n-1 is the largest id in te index).
+        TEST_DATA_T query[d];
+        GenerateVector<TEST_DATA_T>(query, d, n);
+
+        VecSimBatchIterator *batchIterator = VecSimBatchIterator_New(tiered_index, query, nullptr);
+
+        size_t iteration_num = 0;
+        size_t n_res = 1, expected_n_res = 1;
+        while (VecSimBatchIterator_HasNext(batchIterator)) {
+            iteration_num++;
+            // Expect to get results in the reverse order of labels - which is the order of the
+            // distance from the query vector. Get one result in every iteration.
+            auto verify_res = [&](size_t id, double score, size_t index) {
+                ASSERT_EQ(id, iteration_num) << decider_name;
+            };
+            runBatchIteratorSearchTest(batchIterator, n_res, verify_res, BY_SCORE, expected_n_res);
+        }
+
+        ASSERT_EQ(iteration_num, n_labels) << decider_name;
+        VecSimBatchIterator_Free(batchIterator);
+    }
+}
+
+TYPED_TEST(HNSWTieredIndexTest, BatchIteratorAdvanced) {
+    size_t d = 4;
+    size_t M = 8;
+    size_t ef = 1000;
+    size_t n = 1000;
+
+    size_t per_label = TypeParam::isMulti() ? 10 : 1;
+    size_t n_labels = n / per_label;
+
+    // Create TieredHNSW index instance with a mock queue.
+    HNSWParams hnsw_params = {
+        .type = TypeParam::get_index_type(),
+        .dim = d,
+        .metric = VecSimMetric_L2,
+        .multi = TypeParam::isMulti(),
+        .initialCapacity = n,
+        .efConstruction = ef,
+    };
+    VecSimParams params = CreateParams(hnsw_params);
+    HNSWRuntimeParams hnswRuntimeParams = {.efRuntime = ef};
+    VecSimQueryParams query_params = CreateQueryParams(hnswRuntimeParams);
+
+    std::shared_ptr<VecSimAllocator> allocator = VecSimAllocator::newVecsimAllocator();
+    // for (auto &[decider_name, decider] : lambdas) { // TODO: not supported by clang < 16
+    for (auto &lambda : lambdas) {
```

Comment:
Why run over all the combinations here?

Reply:
Why not? To cover different scenarios of vectors and results distribution between the internal indexes